### PR TITLE
Ingest ingest then create index

### LINKED ISSUE #36545: If an ingest node changes the target index, the original index is still created.
**Elasticsearch version** elasticsearch >= 6.5

**Plugins installed**: None

**JVM version** (11.0.1)

Date name processor creates prefixed index additionally to the date timed index. 
This comportement has been introduced by this PR that creates the index before executing the pipeline: https://github.com/elastic/elasticsearch/pull/32786

**Steps to reproduce**:

From a clean elasticsearch instance (from 6.5), just follow the example from date name processor documentation: 

https://www.elastic.co/guide/en/elasticsearch/reference/current/date-index-name-processor.html

```sh
# Create the processor
curl -X PUT "localhost:9200/_ingest/pipeline/monthlyindex" -H 'Content-Type: application/json' -d'
{
  "description": "monthly date-time index naming",
  "processors" : [
    {
      "date_index_name" : {
        "field" : "date1",
        "index_name_prefix" : "myindex-",
        "date_rounding" : "M"
      }
    }
  ]
}
'

# Index a document

curl -X PUT "localhost:9200/myindex/_doc/1?pipeline=monthlyindex" -H 'Content-Type: application/json' -d'
{
  "date1" : "2016-04-25T12:02:01.789Z"
}
'


# my_index and myindex-2016-04-01 are created
curl -XGET "http://localhost:9200/_cat/indices?v"

```

### PR DESCRIPTION
**Pull request by jakelandis**
Execute ingest node pipeline before creating the index

Prior to this commit (and after 6.5.0), if an ingest node changes
the _index in a pipeline, the original target index would be created.
For daily indexes this could create an extra, empty index per day.

This commit changes the TransportBulkAction to execute the ingest node
pipeline before attempting to create the index. This ensures that the 
only index created is the original or one set by the ingest node pipeline. 
This was the execution order prior to 6.5.0 (#32786). 

The execution order was changed in 6.5 to better support default pipelines. 
Specifically the execution order was changed to be able to read the settings
from the index meta data. This commit also includes a change in logic such 
that if the target index does not exist when ingest node pipeline runs, it 
will now pull the default pipeline (if one exists) from the settings of the 
best matched of the index template. 

Relates #32786
Relates #32758 
Closes #36545 

--------

There is some history here...

Originally ingest node executed before any creation of indices. We introduced default pipelines and the discovered https://github.com/elastic/elasticsearch/issues/32758 (the first doc in an index failed to get run through the pipeline).  So we fixed that via https://github.com/elastic/elasticsearch/issues/32758 (changes to create the index first). However, that fix introduced https://github.com/elastic/elasticsearch/issues/36545#issuecomment-447951097 (note - the title is misleading) where now if the pipeline re-directs the `_index` it still creates the original `_index` resulting in an empty index. This is particularly problematic for daily indices that always re-directed, since it would result in an empty index per day. This (re-directing daily indices) is also how upgrading monitoring templates work, and thus this PR is a blocker to  https://github.com/elastic/elasticsearch/pull/39336. 

The fix here is to go back to the original order where ingest node executes before creation of indexes, but solve https://github.com/elastic/elasticsearch/issues/32758 differently. Instead of creating the index and always reading the index meta data, if the index does not exist we can now do the index template matching to look for the default pipeline.

### CODE REVIEW

File: modules/ingest-common/src/test/resources/rest-api-spec/test/ingest/220_drop_processor.yml
Comment (jakelandis) [inline modules/ingest-common/src/test/resources/rest-api-spec/test/ingest/220_drop_processor.yml]: This is removed because with this change the `test` index is no longer created since the create index only happens after ingest (and this drops the index request, thus no index ).

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: Moving the ingest node execution before the create index is the primary change of this PR. 

However, to allow the prevent re-introducing https://github.com/elastic/elasticsearch/issues/32758, we need to look up the default pipeline from the matched templates which is what this code block does.

File: server/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java
Comment (jakelandis) [inline server/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java]: This test is a good test, but it's name was slightly misleading (it still passed with my change to run pipeline before the create index)

File: docs/reference/ingest/ingest-node.asciidoc
Comment (jakelandis) [inline docs/reference/ingest/ingest-node.asciidoc]: these were removed because now that the ingest happens before index creation, if the index doesn't exist yet, and we only drop a document, the index never exists, and thus the response is different. There are new tests explicitly for that in the REST tests.

**Comment by jakelandis** [comment]
@martijnvg @original-brownbear thanks for taking a look.

I forgot to pull the PR out of draft mode, could I get a review approval or LGTM (assuming this is good to go).

**Comment by original-brownbear** [comment]
@jakelandis giving it a thorough review now then, won't take long :)

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: This is a little confusing to me, shouldn't we just loop in standard order and break out once we have a pipeline?
We're doing the reverse iteration elsewhere because we merge all the template's settings but for just this one setting this seems wasteful?

**Review by original-brownbear — CHANGES_REQUESTED** [review]
Only one problem, otherwise LGTM

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: The template with the highest order should be used. The `MetaDataIndexTemplateService.findTemplates(...)` method orders the template by order in descending order. So we can just pick the first index template? (even though that template may not have a default pipeline defined)

Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: I need to emulate how the setting is retrieved for the concrete index so that the behavior here is the same on the first document as it as on the remainder.

For example, say that an index is manually created and matched 3 templates, a default pipeline is defined on the second template match. The index gets that set as the default pipeline. If we picked just the first matching template then we risk choosing a different (or none at all) default pipeline for the first document. 

It is parallel logic which isn't ideal, but feel that it relatively safe here. 

For example:
```
PUT _template/one
{
  "index_patterns": "foobar",
  "order": 1
}

PUT _template/two
{
  "index_patterns": "foobar",
  "order": 2,
  "settings": {
    "default_pipeline": "two"
  }
}

PUT _template/three
{
  "index_patterns": "foobar",
  "order": 3
}

PUT foobar
GET foobar
```
results in `"default_pipeline" : "two",` , which will be the pipeline used that we need to match for the very first document.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: But in your example, don't we just loop to the last template here and get settings that contain any default pipeline and set the default (because nothing is set) `IngestService.NOOP_PIPELINE_NAME` which would be wrong?

Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: @original-brownbear - good catch, i think you are right.  I am only test for precedence for templates with the `default_pipeline` defined. I believe that I can first check to see if the setting exists in the template thereby ignoring the default `IngestService.NOOP_PIPELINE_NAME`.  

Will update.

Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: @original-brownbear - updated https://github.com/elastic/elasticsearch/pull/39607/commits/efdaf538a47d7aa0cb289524541fd5adeaf871a9 (and updated test to cover this case)

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: @jakelandis thanks, this looks like it will yield the correct result, but (I fully see the possibility that I"m wrong here :D), why not simply:

```java
for (IndexTemplateMetaData template : templates) {
                        final Settings settings = template.settings();
                        if (IndexSettings.DEFAULT_PIPELINE.exists(settings)) {
                            defaultPipeline = IndexSettings.DEFAULT_PIPELINE.get(settings);
                            break;
                        }
                    }
```
 
this seems to have the exact same effect as iterating in reverse order and simply overwriting a bunch of values until you iterated through the whole list?

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: Thanks @jakelandis for explaining. I see now that simply picking the last index template is wrong.

I think that @original-brownbear suggestion make sense here. A normal loop based on the list returned by `MetaDataIndexTemplateService.findTemplates(...)` and break when the first pipeline is found is clearer then reserve iterating what the list returned by `MetaDataIndexTemplateService.findTemplates(...)`.

**Review by martijnvg — APPROVED** [review]
👍

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: @original-brownbear @martijnvg - thanks that does make sense for a single setting ! (I was bit slow understanding).

Changed loop here: https://github.com/elastic/elasticsearch/pull/39607/commits/80d71b43445535f47173481c4946491b8e50144e

**Review by original-brownbear — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: @jakelandis thanks LGTM :)

**Pull request merged**
